### PR TITLE
[lldb][libc++] Adds valarray data formatters.

### PR DESCRIPTION
**Pull request by mordante**
The code is heavily based on the vector data formatter.

### CODE REVIEW

File: lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp [hunk 1]
```diff
@@ -750,6 +750,11 @@ static void LoadLibCxxFormatters(lldb::TypeCategoryImplSP cpp_category_sp) {
       lldb_private::formatters::LibcxxStdVectorSyntheticFrontEndCreator,
       "libc++ std::vector synthetic children",
       "^std::__[[:alnum:]]+::vector<.+>$", stl_deref_flags, true);
+  AddCXXSynthetic(
+      cpp_category_sp,
+      lldb_private::formatters::LibcxxStdValarraySyntheticFrontEndCreator,
+      "libc++ std::valarray synthetic children",
+      "^std::__[[:alnum:]]+::valarray<.+>(( )?&)?$", stl_deref_flags, true);
```
Comment (Michael137) [inline lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp:757]: Do we need the `(( )?&)?` part?

File: lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp [hunk 2]
```diff
@@ -871,6 +876,10 @@ static void LoadLibCxxFormatters(lldb::TypeCategoryImplSP cpp_category_sp) {
                 lldb_private::formatters::LibcxxContainerSummaryProvider,
                 "libc++ std::vector summary provider",
                 "^std::__[[:alnum:]]+::vector<.+>$", stl_summary_flags, true);
+  AddCXXSummary(
+      cpp_category_sp, lldb_private::formatters::LibcxxContainerSummaryProvider,
+      "libc++ std::valarray summary provider",
+      "^std::__[[:alnum:]]+::valarray<.+>(( )?&)?$", stl_summary_flags, true);
```
Comment (Michael137) [inline lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp:882]: Do we need `(( )?&)?` here?

File: lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  lldb::ValueObjectSP GetChildAtIndex(size_t idx) override;
+
+  bool Update() override;
+
+  bool MightHaveChildren() override;
+
+  size_t GetIndexOfChildWithName(ConstString name) override;
+
+private:
+  ValueObject *m_start = nullptr;
```
Comment (Michael137) [inline lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp:38]: Can we add brief doxygen comments explaining what these represent? At least for `m_start` and `m_finish`

Comment (Michael137) [reply lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp:38]: I think we should also add a comment about lifetime here (e.g., `pointer is held, not owned`)

File: lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+bool lldb_private::formatters::LibcxxStdValarraySyntheticFrontEnd::
+    MightHaveChildren() {
+  return true;
+}
+
+size_t lldb_private::formatters::LibcxxStdValarraySyntheticFrontEnd::
+    GetIndexOfChildWithName(ConstString name) {
+  if (!m_start || !m_finish)
+    return UINT32_MAX;
```
Comment (Michael137) [inline lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp:130]: should we use `std::numeric_limits` here?

**Review by Michael137 — APPROVED** [review]
LGTM to me generally (just some minor stylistic comments)

File: lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp (hunk 1 above)
Comment (kastiglione) [reply lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp:757]: Not needed, and relatedly I removed a bunch of these https://github.com/llvm/llvm-project/pull/79644

File: lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  uint64_t offset = idx * m_element_size;
+  offset = offset + m_start->GetValueAsUnsigned(0);
+  StreamString name;
+  name.Printf("[%" PRIu64 "]", (uint64_t)idx);
+  return CreateValueObjectFromAddress(name.GetString(), offset,
+                                      m_backend.GetExecutionContextRef(),
+                                      m_element_type);
+}
+
+bool lldb_private::formatters::LibcxxStdValarraySyntheticFrontEnd::Update() {
```
Comment (Michael137) [inline lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp:96]: The return type of `Update` changed in https://github.com/llvm/llvm-project/pull/80167. So you'll have some conflicts when rebasing. The `return false;` should be `return ChildCacheState::eRefetch` instead.

File: lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp (hunk 1 above)
Comment (mordante) [reply lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp:757]: I copied these from `vector` which were present the original tree I worked on. I'll remove them.

File: lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp (hunk 5 above)
Comment (mordante) [reply lldb/source/Plugins/Language/CPlusPlus/LibCxxValarray.cpp]: Thanks for the heads up! I didn't have rebase conflicts, but obviously the code no longer compiled.

**Review by Michael137 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**